### PR TITLE
feat: add p/demo/bf (brainfuck vm)

### PR DESCRIPTION
**Pull request by moul**
Signed-off-by: Manfred Touron <94029+moul@users.noreply.github.com>

# Description

Please provide a detailed description of the changes made in this pull request.

## Contributors Checklist

- [x] Added new tests, or not needed, or not feasible
- [x] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [x] Updated the official documentation or not needed
- [x] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [x] Added references to related issues and PRs
- [x] Provided any useful hints for running manual tests

## Maintainers Checklist

- [ ] Checked that the author followed the guidelines in `CONTRIBUTING.md`
- [ ] Checked the conventional-commit (especially PR title and verb, presence of `BREAKING CHANGE:` in the body)
- [ ] Ensured that this PR is not a significant change or confirmed that the review/consideration process was appropriate for the change

### CODE REVIEW

**Comment by thehowl** [comment]
Just gotta make it run https://github.com/matslina/awib to see Gno compile BF to i386 binaries :)

**Comment by moul** [comment]
Next step: EVM.

File: examples/gno.land/p/demo/bf/bf.gno [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+	// Loop through each character in the code
+	for i := 0; i < len(code); i++ {
+		switch code[i] {
+		case '>':
+			// Increment memory pointer
+			pointer++
+		case '<':
+			// Decrement memory pointer
+			pointer--
```
Comment (thehowl) [inline examples/gno.land/p/demo/bf/bf.gno:22]: many brainfuck interpreters [wrap](https://esolangs.org/wiki/Brainfuck#Conventions) on {over,under}flow. right now we would just panic in those conditions, I suggest we add wrapping

File: examples/gno.land/p/demo/bf/bf.gno [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+				braceCount := 1
+				for braceCount > 0 {
+					i++
+					if code[i] == '[' {
+						braceCount++
+					} else if code[i] == ']' {
+						braceCount--
+					}
+				}
+			}
```
Comment (thehowl) [inline examples/gno.land/p/demo/bf/bf.gno:48]: we could make this use a stack of open brackets, popped each time we go out of a loop, probably better for memory/efficiency... but if you don't want to do it it's fine :)

**Review by thehowl — APPROVED** [review]

File: examples/gno.land/p/demo/bf/bf.gno (hunk 1 above)
Comment (moul) [reply examples/gno.land/p/demo/bf/bf.gno]: done, thanks

File: examples/gno.land/p/demo/bf/bf.gno (hunk 2 above)
Comment (moul) [reply examples/gno.land/p/demo/bf/bf.gno:56]: let's keep this for when we'll have better benchmarking tools as a good example of runtime and gas optimization.

**Pull request merged**